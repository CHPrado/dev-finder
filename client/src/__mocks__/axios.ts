const axios = {
  get: jest.fn(() => Promise.resolve({})),
  post: jest.fn(() => Promise.resolve({})),
  create: () => axios,
  defaults: {
    adapter: {},
  },
};

export default axios;
