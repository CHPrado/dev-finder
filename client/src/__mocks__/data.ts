import { createMemoryHistory, createLocation } from "history";

import { DevProps, DevRouteProps } from "../interfaces";

const devData: DevProps = {
  username: "CHPrado",
  name: "Christian Prado",
  bio: "My bio",
  avatar_url: "avatar_url",
  location: "Lins - SP",
  followers: 1,
  following: 2,
  public_repos: 3,
  stars: 4,
  watchers: 5,
  forks: 6,
  languages: [
    {
      name: "JavaScript",
      count: 1,
      percent: "49.9",
    },
    {
      name: "Ruby",
      count: 1,
      percent: "51.1",
    },
  ],
  notes: "Notes about me",
};

const devRouteData: DevRouteProps = {
  history: createMemoryHistory(),
  location: createLocation({ state: { dev: devData } }),
  match: {
    isExact: false,
    path: "/dev",
    url: "/dev",
    params: {},
  },
};

export { devData, devRouteData };
