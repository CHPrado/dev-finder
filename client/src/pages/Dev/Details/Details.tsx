import React, { FC } from "react";
import { Link } from "react-router-dom";

import { Box, Typography } from "@material-ui/core";
import {
  RepoIcon,
  StarIcon,
  PeopleIcon,
  GitForkIcon,
  EyeIcon,
} from "@primer/octicons-react";

import { DevProps } from "../../../interfaces";
import useStyles from "./styles";

interface Props {
  dev: DevProps;
}

const Details: FC<Props> = ({ dev }) => {
  const classes = useStyles();

  return (
    <Box className={classes.root}>
      <Box width="100%">
        <Typography className={classes.value}>{dev.followers}</Typography>
        <Typography className={classes.value}>{dev.following}</Typography>
        <Typography className={classes.value}>{dev.public_repos}</Typography>
        <Typography className={classes.value}>{dev.stars}</Typography>
        <Typography className={classes.value}>{dev.watchers}</Typography>
        <Typography className={classes.value}>{dev.forks}</Typography>
      </Box>

      <Box display="flex" flexDirection="column" width="100%">
        <Link
          className={classes.link}
          to={(location) => ({
            ...location,
            pathname: `http://www.github.com/${dev.username}?tab=followers`,
          })}
          target="_blank"
        >
          <Typography className={classes.text}>
            <PeopleIcon size={16} className={classes.icon} />
            followers
          </Typography>
        </Link>
        <Link
          className={classes.link}
          to={(location) => ({
            ...location,
            pathname: `http://www.github.com/${dev.username}?tab=following`,
          })}
          target="_blank"
        >
          <Typography className={classes.text}>
            <PeopleIcon size={16} className={classes.icon} />
            following
          </Typography>
        </Link>
        <Link
          className={classes.link}
          to={(location) => ({
            ...location,
            pathname: `http://www.github.com/${dev.username}?tab=repositories`,
          })}
          target="_blank"
        >
          <Typography className={classes.text}>
            <RepoIcon size={16} className={classes.icon} />
            public repositories
          </Typography>
        </Link>
        <Link
          className={classes.link}
          to={(location) => ({
            ...location,
            pathname: `http://www.github.com/${dev.username}?tab=stars`,
          })}
          target="_blank"
        >
          <Typography className={classes.text}>
            <StarIcon size={16} className={classes.icon} />
            stars
          </Typography>
        </Link>
        <Typography className={classes.text}>
          <EyeIcon size={16} className={classes.icon} />
          watchers
        </Typography>
        <Typography className={classes.text}>
          <GitForkIcon size={16} className={classes.icon} />
          forks
        </Typography>
      </Box>
    </Box>
  );
};

export default Details;
