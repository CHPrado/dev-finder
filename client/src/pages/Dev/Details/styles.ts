import { Theme, makeStyles } from "@material-ui/core/styles";

export default makeStyles((theme: Theme) => ({
  root: {
    display: "flex",
    marginRight: theme.spacing(2),
    [theme.breakpoints.down("xs")]: {
      margin: 0,
    },
  },
  value: {
    fontWeight: theme.typography.fontWeightBold,
    color: theme.palette.text.secondary,
    margin: theme.spacing(0, 1),
    textAlign: "right",
    display: "flex",
    alignItems: "center",
    justifyContent: "flex-end",
  },
  link: {
    textDecoration: "none",
    "& :hover": {
      color: theme.palette.primary.main,
      textDecoration: "underline",
    },
  },
  text: {
    color: theme.palette.text.secondary,
    margin: theme.spacing(0, 1),
    whiteSpace: "nowrap",
  },
  icon: {
    marginRight: theme.spacing(2),
  },
}));
