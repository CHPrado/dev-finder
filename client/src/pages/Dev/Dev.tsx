import React, { FC, useState } from "react";
import { Link } from "react-router-dom";

import {
  Box,
  Card,
  CircularProgress,
  IconButton,
  Tooltip,
  Typography,
} from "@material-ui/core";
import { CachedOutlined } from "@material-ui/icons";
import { LocationIcon } from "@primer/octicons-react";

import { Header, Snackbar } from "../../components";
import { SnackbarProps, DevRouteProps } from "../../interfaces";
import { githubApi } from "../../services";
import { Details } from "./Details";
import { Languages } from "./Languages";
import { Notes } from "./Notes";
import useStyles from "./styles";

const Dev: FC<DevRouteProps> = (props) => {
  const classes = useStyles();
  const [dev, setDev] = useState(props.location.state.dev);
  const [isLoading, setIsLoading] = useState(false);
  const [snackbar, setSnackbar] = useState<SnackbarProps>({});

  const handleRefresh = () => {
    setIsLoading(true);
    githubApi
      .getDev(dev.username)
      .then((resp) => {
        setDev(resp);
      })
      .catch(() => {
        setSnackbar({
          open: true,
          message:
            "There was an error trying to refresh the data. Please, try again.",
          type: "error",
          onClose: () => setSnackbar({ open: false }),
        });
      })
      .finally(() => setIsLoading(false));
  };

  return (
    <Box className={classes.root}>
      <Header />

      {isLoading ? (
        <Box className={classes.loadingBox}>
          <CircularProgress size="100px" />
        </Box>
      ) : (
        <Box className={classes.container}>
          <Card variant="outlined" className={classes.card}>
            <Box className={classes.titleBox}>
              <Link
                className={classes.link}
                to={(location) => ({
                  ...location,
                  pathname: `http://www.github.com/${dev.username}`,
                })}
                target="_blank"
              >
                <Typography className={classes.name}>
                  {`${dev.username}'s profile`}
                </Typography>
              </Link>

              <Box display="flex">
                <Tooltip title="Refresh">
                  <IconButton onClick={handleRefresh}>
                    <CachedOutlined />
                  </IconButton>
                </Tooltip>
              </Box>
            </Box>

            <Box className={classes.avatarBox}>
              <img
                src={dev.avatar_url}
                alt="avatar"
                className={classes.avatar}
              />

              <Box className={classes.infoBox}>
                <Typography className={classes.info}>{dev.name}</Typography>
                {dev.location && (
                  <Typography className={classes.info}>
                    <LocationIcon size={18} className={classes.icon} />
                    {dev.location}
                  </Typography>
                )}
                <Typography color="textSecondary" align="justify">
                  {dev.bio}
                </Typography>
              </Box>
            </Box>

            <Box className={classes.detailsBox}>
              <Details dev={dev} />
              <Languages languages={dev.languages} />
            </Box>

            <Notes
              dev={dev}
              setIsLoading={setIsLoading}
              setSnackbar={setSnackbar}
              setDev={setDev}
            />
          </Card>
        </Box>
      )}

      <Snackbar {...snackbar} />
    </Box>
  );
};

export default Dev;
