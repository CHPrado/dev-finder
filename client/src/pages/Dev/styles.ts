import { Theme, makeStyles } from "@material-ui/core/styles";

const flexCenter = {
  display: "flex",
  alignItems: "center",
  justifyContent: "center",
};

export default makeStyles((theme: Theme) => ({
  root: {
    display: "flex",
    flexDirection: "column",
    height: "100%",
  },
  loadingBox: {
    height: "100%",
    ...flexCenter,
  },
  container: {
    ...flexCenter,
    position: "absolute",
    top: "100px",
    width: "100%",
  },
  card: {
    ...flexCenter,
    flexDirection: "column",
    margin: theme.spacing(5),
    padding: theme.spacing(5),
    maxWidth: "500px",
    [theme.breakpoints.down("xs")]: {
      width: "100%",
    },
  },
  titleBox: {
    width: "100%",
    display: "flex",
    alignItems: "center",
    justifyContent: "space-between",
    marginBottom: theme.spacing(4),
  },
  name: {
    color: theme.palette.primary.dark,
    fontSize: theme.typography.h6.fontSize,
    fontWeight: theme.typography.fontWeightBold,
    whiteSpace: "nowrap",
  },
  avatarBox: {
    display: "flex",
    alignItems: "flex-start",
    justifyContent: "center",
    [theme.breakpoints.down("xs")]: {
      flexDirection: "column",
      alignItems: "center",
    },
  },
  avatar: {
    borderRadius: "50%",
    height: "150px",
    marginRight: theme.spacing(4),
    [theme.breakpoints.down("xs")]: {
      height: "200px",
      marginBottom: theme.spacing(4),
    },
  },
  infoBox: {
    marginBottom: theme.spacing(4),
    display: "flex",
    flexDirection: "column",
    width: "100%",
  },
  info: {
    color: theme.palette.text.secondary,
    fontSize: theme.typography.h6.fontSize,
    fontWeight: theme.typography.fontWeightBold,
    display: "flex",
    alignItems: "center",
  },
  detailsBox: {
    width: "100%",
    display: "flex",
    justifyContent: "center",
    margin: theme.spacing(4),
    [theme.breakpoints.down("xs")]: {
      flexDirection: "column",
    },
  },
  icon: {
    marginRight: theme.spacing(2),
  },
  link: {
    textDecoration: "none",
    "& :hover": {
      color: theme.palette.primary.main,
      textDecoration: "underline",
    },
  },
}));
