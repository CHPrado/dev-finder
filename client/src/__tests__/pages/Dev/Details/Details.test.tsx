import React from "react";
import { BrowserRouter } from "react-router-dom";

import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";

import { devData } from "../../../../__mocks__/data";
import { Details } from "../../../../pages/Dev/Details";

const wrapper = mount(
  <BrowserRouter>
    <Details dev={devData} />
  </BrowserRouter>
);

describe("rendering component", () => {
  it("renders Details component without crashing", () => {
    expect(wrapper.contains("followers")).toEqual(true);
    expect(wrapper.contains("following")).toEqual(true);
    expect(wrapper.contains("public repositories")).toEqual(true);
    expect(wrapper.contains("stars")).toEqual(true);
    expect(wrapper.contains("watchers")).toEqual(true);
    expect(wrapper.contains("forks")).toEqual(true);
  });
});

describe("passing props", () => {
  it("dev props data equals to dev data", () => {
    const details = wrapper.find(Details).at(0);

    expect(details.props().dev).toEqual(devData);
  });
});

describe("snapshots", () => {
  it("Details snapshots", () => {
    const tree = shallow(<Details dev={devData} />);

    expect(toJson(tree)).toMatchSnapshot();
  });
});
