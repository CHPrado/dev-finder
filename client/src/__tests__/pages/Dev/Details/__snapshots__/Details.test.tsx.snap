// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots Details snapshots 1`] = `
<Styled(MuiBox)
  className="makeStyles-root-1"
>
  <Styled(MuiBox)
    width="100%"
  >
    <WithStyles(ForwardRef(Typography))
      className="makeStyles-value-2"
    >
      1
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      className="makeStyles-value-2"
    >
      2
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      className="makeStyles-value-2"
    >
      3
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      className="makeStyles-value-2"
    >
      4
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      className="makeStyles-value-2"
    >
      5
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      className="makeStyles-value-2"
    >
      6
    </WithStyles(ForwardRef(Typography))>
  </Styled(MuiBox)>
  <Styled(MuiBox)
    display="flex"
    flexDirection="column"
    width="100%"
  >
    <Link
      className="makeStyles-link-3"
      target="_blank"
      to={[Function]}
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-text-4"
      >
        <PeopleIcon
          className="makeStyles-icon-5"
          size={16}
          verticalAlign="text-bottom"
        />
        followers
      </WithStyles(ForwardRef(Typography))>
    </Link>
    <Link
      className="makeStyles-link-3"
      target="_blank"
      to={[Function]}
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-text-4"
      >
        <PeopleIcon
          className="makeStyles-icon-5"
          size={16}
          verticalAlign="text-bottom"
        />
        following
      </WithStyles(ForwardRef(Typography))>
    </Link>
    <Link
      className="makeStyles-link-3"
      target="_blank"
      to={[Function]}
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-text-4"
      >
        <RepoIcon
          className="makeStyles-icon-5"
          size={16}
          verticalAlign="text-bottom"
        />
        public repositories
      </WithStyles(ForwardRef(Typography))>
    </Link>
    <Link
      className="makeStyles-link-3"
      target="_blank"
      to={[Function]}
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-text-4"
      >
        <StarIcon
          className="makeStyles-icon-5"
          size={16}
          verticalAlign="text-bottom"
        />
        stars
      </WithStyles(ForwardRef(Typography))>
    </Link>
    <WithStyles(ForwardRef(Typography))
      className="makeStyles-text-4"
    >
      <EyeIcon
        className="makeStyles-icon-5"
        size={16}
        verticalAlign="text-bottom"
      />
      watchers
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      className="makeStyles-text-4"
    >
      <GitForkIcon
        className="makeStyles-icon-5"
        size={16}
        verticalAlign="text-bottom"
      />
      forks
    </WithStyles(ForwardRef(Typography))>
  </Styled(MuiBox)>
</Styled(MuiBox)>
`;
